src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,196 @@
+    }
+
+    public void printLinkedList() {
+        Node currentNode = head;
+
+        while (currentNode != null) {
+            System.out.println(currentNode.data);
+            currentNode = currentNode.next;
+        }
+    }
+
+    public void printReversedLinkedList() {
+        Node currentNode = tail;
+
+        while (currentNode != null) {
+            System.out.println(currentNode.data);
+            currentNode = currentNode.prev;
+        }
+    }
+
+    private Node unlink(Node node) {
+        size--;
+        if (node == tail) {
+            tail = node.prev;
+            return node;
+        }
+        if (node == head) {
+            head = node.next;
+            return node;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        return node;
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndexExclusive(index);
+        Node newNode = new Node(value);
+        if (isEmpty()) {
+            head = newNode;
+            tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node currentNode = getNode(index);
+            insertBefore(currentNode, newNode);
+        }
+        size++;
+        System.out.println("------------");
+        printLinkedList();
+        System.out.println("------------");
+        printReversedLinkedList();
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node nodeToUpdate = getNode(index);
+        T oldValue = nodeToUpdate.data;
+        nodeToUpdate.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node nodeToRemove = getNode(index);
+        return unlink(nodeToRemove).data;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = getIndexOfNode(object);
+        if (index == -1) {
+            return false;
+        }
+        remove(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    class Node {
+        private T data;
+        private Node next;
+        private Node prev;
+
+        private Node(T data) {
+            this.data = data;
+        }
```

Comment:
see difference between static-inner class and non-static inner class. (known as nested or inner classes)

And chose the better for our case
```suggestion
    class Node {
        private T data;
        private Node next;
        private Node prev;

        private Node(T data) {
            this.data = data;
        }
```

Reply:
nested classes provide more encapsulation. In this case inner class cannot reference non-static fields from outer class, espesially we cannot use parametrized value from class, should create our own

Reply:
better i suppose is static, it gives more control, but works a bit slowly becouse of asdditional parametrisation